zebra/zserv.c

```diff
@@ -587,18 +587,12 @@ static void zserv_client_free(struct zserv *client)
 
 		close(client->sock);
 
-		/* If this is a synchronous BGP Zebra client for label/table
-		 * manager, then ignore it. It's not GR-aware, and causes GR to
-		 * be skipped for the session_id == 0 (asynchronous).
-		 */
-		if (client->proto == ZEBRA_ROUTE_BGP && client->session_id == 1)
-			return;
-
 		if (DYNAMIC_CLIENT_GR_DISABLED(client)) {
 			zebra_mpls_client_cleanup_vrf_label(client->proto);
 
 			nroutes = rib_score_proto(client->proto,
-						  client->instance);
+						  client->instance,
+						  client->session_id);
```

Comment:
hmm, could we look at the actual client, and see that it's synchronous? I don't love the bgp-specific test (I guess)

Reply:
> How can I do that (distinguish if it's an synchronous or not)? :) Introduce an additional flag/variable for zserv?

oh - we have a 'synchronous' bit in the struct zclient and in the zebra struct zserv - you can see it being set in zapi_msg.c:zread_hello() and used in several places in zebra/zapi_msg.c:

```
	for (ALL_LIST_ELEMENTS(zrouter.client_list, node, nnode, client)) {
		/* Do not send unsolicited messages to synchronous clients. */
		if (client->synchronous)
			continue;

		zsend_capabilities(client, zvrf);
	}
```

Reply:
Ah, right, I forgot I was did the same previously :smile:

Reply:
@mjstapp updated, thanks!